### PR TITLE
Make proper test integration for the metrics api

### PR DESCRIPTION
**Pull request by purbon**
Make usage of proper test integration with the main core pipeline for the logstash web api, this basically means this PR is cleaning up all unnecessary mocks and going through a real agent and pipeline for testing purposes.

Still room to improve the test, this this is a first pass, next steps are to actually improve the api endpoint test for expected behaviour.

Small set of test improvements, like:

* Cleanup non used fixture.
* Improve node_stats tests by adding more context and format checks.
* Add a new node_specs endpoint test that checks the expected data is
  returned for the hot_threads resource.

@ph what do you think?

//cc @jsvd

### CODE REVIEW

File: logstash-core/lib/logstash/api/lib/app/service.rb [hunk 1]
```diff
@@ -40,4 +44,15 @@ def get(key)
     end
     LogStash::Json.dump(data)
   end
+
+  private
+
+  def has_counters?
```
Comment (purbon) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: This was actually a source of concern to me, as for testing I do really need to wait till some events are generated. Obviously open to improve this part of the code.

File: rakelib/default_plugins.rb [hunk 2]
```diff
@@ -119,6 +119,8 @@ module RakeLib
       logstash-input-stdin
       logstash-input-tcp
       logstash-output-stdout
+      logstash-output-null
```
Comment (suyograo) [inline rakelib/default_plugins.rb]: why this change? Can't we add this to dev dependencies?

Comment (purbon) [inline rakelib/default_plugins.rb]: we could yes, the idea is to use this in the agent pipeline for this test, so nothing is really output, but still everything is flowing for all the stages.

File: logstash-core/lib/logstash/api/lib/app/service.rb (hunk 1 above)
Comment (ph) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: What I did in my test is to use the generator and custom output with an accessor for the number of events received.
So I wait until the output has received all the events generated by the generator.

File: logstash-core/lib/logstash/api/lib/app/service.rb [hunk 3]
```diff
@@ -40,4 +44,15 @@ def get(key)
...
+    return false if @snapshot.nil?
+    if @snapshot_rotation_mutex.try_lock
+      has = @snapshot.metric_store.all.clone.map { |t| t.class.to_s }.include?("LogStash::Instrument::MetricType::Counter")
+      @snapshot_rotation_mutex.unlock
+      return has
+    end
+  end
```
Comment (ph) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: Since I have the `@fast_lookup` now I can add a `size` or has_metrics? Method would that help you?

Comment (ph) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: Also since all metrics are lazy created if you do `@snapshot.metric_store.all.size > 0` this will tell you if we have any metrics in the store.  

But lets create a predicated for it. `#empty?` `#present?` Naming is hard :)

File: logstash-core/spec/api/lib/api/node_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    end
+
+    context "when asking for human output" do
+      before(:all) do
+        get "/hot_threads?human"
+      end
+
+      let(:payload) { last_response.body }
+
+      it "shuold return a plain text description" do
```
Comment (ph) [inline logstash-core/spec/api/lib/api/node_spec.rb]: small typo `shuold` -> `should`, Should we check for `content_type`?

File: logstash-core/spec/api/spec_helper.rb [hunk 5]
```diff
@@ -17,3 +17,40 @@ def read_fixture(name)
...
+    allow(LogStash::Instrument::Collector.instance).to receive(:agent).and_return(agent)
+  end
+end
+
+def wait_until_snapshot_received
+  while !LogStash::Api::Service.instance.started? do
+    sleep 0.5
+  end
+end
+wait_until_snapshot_received
```
Comment (ph) [inline logstash-core/spec/api/spec_helper.rb]: I have a small question could this `spec_helper` make some conflict with our other `spec/spec_helper` ?

**Comment by ph** [comment]
@purbon I really like that! thanks for making it happen. I have similar test for the pipeline in #4670

File: logstash-core/lib/logstash/api/lib/app/service.rb (hunk 3 above)
Comment (purbon) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: that would help, but my issue here is more on the side of having counters there. it takes some time for the pipeline to actually generate some events acouting information, and to test for example the  `/events` resource I do need this to be there. JVM periodic pollers is working like charm and providing data very fast.

File: logstash-core/lib/logstash/api/lib/app/service.rb (hunk 1 above)
Comment (purbon) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: That could be a good idea, but I'm more on the side of letting the collector notify when he has received events, or any other side of information we aim for. On the other side, I'm starting to believe we should cleanup all this real/dummy plugins and move to a unified solution here.

Comment (ph) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: Well, we can also change the behavior if the collector didn't collect any new metrics we dont send the snapshot?

File: logstash-core/lib/logstash/api/lib/app/service.rb (hunk 3 above)
Comment (ph) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: true, if you are targetting the events.

File: rakelib/default_plugins.rb (hunk 2 above)
Comment (purbon) [inline rakelib/default_plugins.rb]: Sorry to say, we could not as we need to have this as runtime dependencies because of the plugins system.

File: logstash-core/lib/logstash/api/lib/app/service.rb (hunk 1 above)
Comment (purbon) [inline logstash-core/lib/logstash/api/lib/app/service.rb]: well, it collects metrics somehow, as it collect the periodic poller metrics anyway. this is only a testing issue for now.

File: logstash-core/spec/api/spec_helper.rb (hunk 5 above)
Comment (purbon) [inline logstash-core/spec/api/spec_helper.rb]: I don't think so as is only used by the API tests and is only locally required by them.

**Comment by purbon** [comment]
closing this in benefit of a new PR to master, as the feature/metrics branch got merged recently.
